db/database_test.go

```diff
@@ -1907,3 +1908,179 @@ func TestSyncFnMutateBody(t *testing.T) {
 	log.Printf("rev: %s", rev.BodyBytes)
 
 }
+
+// Multiple clients are attempting to push the same new revision concurrently; first writer should be successful,
+// subsequent writers should fail on CAS, and then identify that revision already exists on retry.
+func TestConcurrentPushSameNewRevision(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelDebug, base.KeyAll)()
+	db, testBucket := setupTestDB(t)
+	defer testBucket.Close()
+	defer tearDownTestDB(t, db)
+
+	success, failure := int64(0), int64(0)
+	var wg sync.WaitGroup
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			body := Body{"name": "Bob", "age": 52}
+			revId, doc, err := db.Put("doc1", body)
+			if err != nil {
+				assert.Equal(t, "409 Document exists", err.Error())
+				atomic.AddInt64(&failure, 1)
+			} else {
+				atomic.AddInt64(&success, 1)
+				assert.NotEmpty(t, revId)
+				assert.Equal(t, "Bob", doc._body["name"])
+				assert.Equal(t, 52, doc._body["age"])
+			}
+		}()
+	}
+	wg.Wait()
+	assert.Equal(t, int64(1), atomic.LoadInt64(&success))
+	assert.Equal(t, int64(4), atomic.LoadInt64(&failure))
+}
+
+// Multiple clients are attempting to push the same new, non-winning revision concurrently; non-winning is an
+// update to a non-winning branch that leaves the branch still non-winning (i.e. shorter) than the active branch
+func TestConcurrentPushSameNewNonWinningRevision(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelDebug, base.KeyAll)()
+	db, testBucket := setupTestDB(t)
+	defer testBucket.Close()
+	defer tearDownTestDB(t, db)
+
+	body := Body{"name": "Olivia", "age": 80}
+	_, _, err := db.PutExistingRevWithBody("doc1", body, []string{"1-a"}, false)
+	assert.NoError(t, err, "Adding revision 1-a")
+
+	body = Body{"name": "Harry", "age": 40}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 2-a")
+
+	body = Body{"name": "Amelia", "age": 20}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"3-a", "2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 3-a")
+
+	body = Body{"name": "Charlie", "age": 10}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"4-a", "3-a", "2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 4-a")
+
+	body = Body{"name": "Noah", "age": 40}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"2-b", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 2-b")
+
+	var wg sync.WaitGroup
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			body := Body{"name": "Emily", "age": 20}
+			_, _, err := db.PutExistingRevWithBody("doc1", body, []string{"3-b", "2-b", "1-a"}, false)
+			assert.NoError(t, err, "Adding revision 3-b")
```

Comment:
I need to dig into `PutExistingRev` a little more to understand it in this (and below) scenario, because I expected something else to happen.

My expectation was that only one of these would succeed. I'm not sure that's the case if all of these requests work without producing an error.

Reply:
For all but the first goroutine, I'd expect PutExistingRevWithBody to identify that it already has all the incoming revisions, and so ignore without returning an error (via the ErrUpdateCancel handling in updateAndReturnDoc)